### DISCUSSION TITLE
Getting data from Home Assistant into Tasmota

### DISCUSSION THREAD

**Discussion by richardsg307** [arendst/Tasmota#20233] [Q&A] [answered]
I'm on a project to replace an ESP8266, with own code that monitors my boiler and reports to HA, with an ESP32 + display. I've gone for the Lilygo T-Display-S3 and got it running with Tasmota, Berry, LVGL and HASPmota. On the breadboard I have a couple of DS18x20 and a button to simulate the Burning sensor I have. These I can now display OK. I've also worked out (with help on this forum) how to get variables in Berry onto the display. Here's a pic of what I've got so far.
![20231214_004506](https://github.com/arendst/Tasmota/assets/67514113/18217151-7260-4ca6-b8d1-d283697c0f7e)
I now need some help to work out how to get data from Home Assistant into Tasmota and variables in Berry (ie the numbers down the right hand column on the screen). MQTT or Commands seem to be options; I know how to issue either from HA but it's the receiving Tasmota/Berry side I can't work out from the docs. There's also a suggestion that I might need to compile my own Tasmota to achieve MQTT subscribing! And that fills me with dread.
If I get it cracked I'll post full details here so others will benefit.

**Comment by Noschvie** [comment]
„self compiled“ can easily be done using Tasmota online compiler or using the Docker image.
 https://tasmota.github.io/docs/Compile-your-build/

**Reply by barbudor** [reply]
MQTT subscription in Berry works out of the box. No self compile is required.

You may have read information about USE_MQTT_EVENT build flag which only applies at Rules and Scripting-Language

AFAIK, there are no optional features in Berry, you always get the full cake and topings

**Reply by sfromis** [reply]
Assuming that you have control over what you're publishing from HA, you do not really need to manage it as a subscription. You could also register a Tasmota command name with Berry, that way it would be available from HA by publishing a Tasmota command the normal way, but also available to testing from the Tasmota console, or even other automations. 

As I already mentioned [in the other thread](https://github.com/arendst/Tasmota/discussions/20228#discussioncomment-7842130), another approach could be to use the existing `Br` command to execute a snippet of Berry code to do the work.

(There are a few specialized things in Berry which can be enabled build time, but this is outside of "standard usage" like subscriptions or commands).

**Reply by richardsg307** [reply]
I've looked through the Berry documentation and can find no mention of Br nor vt coding. Berry, LVGL and HASPmota are all new to me and I learn best by seeing examples or real code rather than the sometimes cryptic documentation. I'll try the Tasmota command/MQTT subscribe approaches later today or a TCP/UDP method perhaps, which I have read in the Berry docs and am familiar with at the HA (Node-Red) end.

**Reply by sfromis** [reply]
The `Br` command is documented on the page for commands, including an example.
https://tasmota.github.io/docs/Commands/#br
As shown, the argument is simply a bit of Berry coding, does not have to be as much as in the example. The example I already mentioned in the other thread is what could work with the `Br` command.

And **any** command can be invoked via a MQTT publish.
https://tasmota.github.io/docs/MQTT/#commands-over-mqtt

Adding a new command name is `tasmota.add_cmd`
https://tasmota.github.io/docs/Berry/#functions-to-create-custom-tasmota-command
An example here:
https://github.com/arendst/Tasmota/blob/development/tasmota/berry/modules/antiburn/autoexec.be

Using the subscribe feature is here:
https://tasmota.github.io/docs/Berry/#mqtt-module
With Berry, the only thing you "gain" using Subscribe is the opportunity to choose a different topic, instead of simply using a command.

I'd not really recommend using subscribe for implementing commands; this approach was made for ESP8266 and subscribing to things like sensor data from other Tasmota devices. While doing such is much more straightforward with Berry, it is not "better" than creating Tasmota commands for your desired new commands.

**Reply by richardsg307** [reply]
Well my first attempt at subscribing bricked the device (constant reboot) and I literally had to reinstall and configure everything from scratch! I'll look through the links and try something else. Thanks.

**Reply by richardsg307** [reply]
I've managed to create a command that works. this is my autoexec
```
import haspmota
haspmota.start()
tasmota.add_cmd("GetStats",
  def ()
      print("Incoming")
      tasmota.resp_cmnd_done()
  end
)
```
Now I've been looking for how to get at the payload of the command and found nothing. Payload could be just a number if I create a different command for each attribute I want to send, or a JSON string containing all attributes. But how do I get at it?
If I use MQTT subscribe I can see that there is an mqtt_data driver that I think would give me the payload but no example of how to use it.

**Reply by sfromis** [reply]
```
tasmota.add_cmd('Test',
  def (cmd, idx, payload, raw)
    print(
      cmd,     # command name
      idx,     # index glued on to command name, or 0
      payload, # payload as string
      data)    # If payload was in JSON format, a map with the parsed data
    tasmota.resp_cmnd_done()
  end)
```

Basically very similar to using `mqtt.subscribe` (where the intention is to use the functions pointed to by my link, and **not** use a driver or `mqtt_data` - which has been replaced by the simpler approach).

**Reply by richardsg307** [reply]
Thanks, that's what I needed. It didn't work as you wrote it ('data' undeclared (first use in this function)) so i chopped that bit of and it worked. I get a 1 for idx and payload is whatever I type after 'Test '. I can work with that now.
I would never have got there from the documentation. I'll get it all set up now and post the completed solution.
Edit: Just spotted 'data' vs 'raw'. Yours does work if I make them both 'raw'.

**Reply by sfromis** [reply]
Yeah, I changed from `raw` to `data` as it does make sense to name a map as "raw", but was a bit too quick.

**Comment by richardsg307** [comment, marked as answer]
I've got it totally working on my breadboard with my old ESP8266 still attached to the boiler and reporting to Home Assistant. So, as promised here's the solution.
The autoexec.be
```
import haspmota
import string
haspmota.start()

tasmota.add_cmd('BoilerStats',
  def (cmd, idx, payload, raw)
    print(
      cmd,     # command name
      idx,     # index glued on to command name, or 0
      payload,  # payload as string
      raw)     # If payload was in JSON format, a map with the parsed data
    global.p1b35.text = string.format("%5.2fHrs",raw["Hours"])
    global.p1b36.text = string.format("%5.2fL",raw["Oil"])
    global.p1b37.text = string.format("%5.2fKWh",raw["KWh"])
    global.p1b38.text = string.format("£%5.2f",raw["Cost"])
    tasmota.resp_cmnd_done()
  end)
```
the print command is not needed now but was useful seeing what was going on via the Berry console. I originally used payload to extract the values but it didn't work (it did on the console!); then I found raw did.
Here's the pages.jsonl file
```
{"page":0,"comment":"---------- Upper stat line ----------"}

{"id":11,"obj":"label","x":0,"y":0,"w":320,"pad_right":90,"h":22,"bg_color":"#D00000","bg_opa":255,"text_color":"#ffffff","radius":0,"border_side":0,"text":"Boiler Stats","text_font":"montserrat-20"}

{"id":15,"obj":"lv_wifi_arcs","x":291,"y":0,"w":29,"h":22,"radius":0,"border_side":0,"bg_color":"#000000","line_color":"#FFFFFF"}
{"id":16,"obj":"lv_clock","x":232,"y":3,"w":55,"h":16,"radius":0,"border_side":0,"text_color":"#ffffff"}

{"page":1,"comment":"---------- Page 1 ----------"}
{"id":0,"bg_color":"#0000A0","bg_grad_color":"#000000","bg_grad_dir":1,"text_color":"#FFFFFF"}

{"id":19,"obj":"label","x":10,"y":30,"w":100,"text":"Local Readings","align":2}
{"id":20,"obj":"label","x":10,"y":50,"w":70,"text":"Outflow = ","align":2}
{"id":21,"obj":"label","x":10,"y":70,"w":70,"text":"Inflow = ","align":2}
{"id":22,"obj":"label","x":10,"y":90,"w":70,"text":"Cabinet = ","align":2}
{"id":23,"obj":"label","x":10,"y":110,"w":70,"text":"Demand = ","align":2}
{"id":24,"obj":"label","x":10,"y":130,"w":70,"text":"Burn = ","align":2}

{"id":25,"obj":"label","x":80,"y":50,"w":60,"align":0,"text_rule":"DS18B20-1#Temperature","text_rule_format":"%4.1f°C"}
{"id":26,"obj":"label","x":80,"y":70,"w":60,"align":0,"text_rule":"DS18B20-2#Temperature","text_rule_format":"%4.1f°C"}
{"id":27,"obj":"label","x":80,"y":90,"w":60,"align":0,"text_rule":"DS18B20-1#Temperature","text_rule_format":"%4.1f°C"}
{"id":28,"obj":"label","x":80,"y":110,"w":60,"align":0,"text_rule":"Switch1#Action","text_rule_format":"%s"}
{"id":29,"obj":"label","x":80,"y":130,"w":60,"align":0,"text_rule":"Switch1#Action","text_rule_format":"%s"}

{"id":30,"obj":"label","x":150,"y":30,"w":100,"text":"HA Stats","align":2}
{"id":31,"obj":"label","x":150,"y":50,"w":80,"text":"Run Time = ","align":2}
{"id":32,"obj":"label","x":150,"y":70,"w":80,"text":"Oil = ","align":2}
{"id":33,"obj":"label","x":150,"y":90,"w":80,"text":"Energy = ","align":2}
{"id":34,"obj":"label","x":150,"y":110,"w":80,"text":"Cost = ","align":2}

{"id":35,"obj":"label","x":230,"y":50,"w":90,"align":0,"text":"Waiting"}
{"id":36,"obj":"label","x":230,"y":70,"w":90,"align":0,"text":"  for  "}
{"id":37,"obj":"label","x":230,"y":90,"w":90,"align":0,"text":"  next "}
{"id":38,"obj":"label","x":230,"y":110,"w":90,"align":0,"text":"report"}
```
I've seperated the labels from the values. Some values get populated from Tasmota and its sensors direct, whilst the others from values calculated and sent by Home Assistant. The yaml for the automation that does the sending every 5 minutes is
```
alias: Boiler Report Stats
description: ""
trigger:
  - platform: time_pattern
    minutes: /5
condition: []
action:
  - service: mqtt.publish
    data:
      qos: 0
      retain: false
      topic: cmnd/boilerdisplay/BoilerStats
      payload_template: |-
        {{"{"+"Hours"|tojson +":"+ states("sensor.burn_time_by_day")+","+ 
        "Oil"|tojson +":"+ states("sensor.oil_litres_today")+","+
        "KWh"|tojson +":"+ states("sensor.oil_energy")+","+
        "Cost"|tojson +":"+ states("sensor.oil_cost_today")+"}"}}
mode: single
```
Note that 'BoilerStats' is the name of the command added to Tasmota in the autoexec.be file, and 'boilerdisplay' is the Topic declared in the Config MQTT page within Tasmota.
And that's it. Simples. I will tweek this a bit when I actually install it but this should help others do similar.
This is the result

![20231214_213935](https://github.com/arendst/Tasmota/assets/67514113/171ee738-ede6-431a-ac80-5ad11292bad0)

Big shout out to sfromis without whose input I would never have achieved this.
